### DISCUSSION TITLE
time to move to notcurses backend?

### DISCUSSION THREAD

**Discussion by dankamongmen** [libsixel/libsixel#57] [Ideas]
*see my comment below; i think maybe @hpjansson's libchafa would be a better backend for this kind of replacement.
--@dankamongmen 2022-02-05*

@ctrlcctrlv, i guess this is most reasonably targeted at you, but i figured i'd make the discussion public.

i've been comaintaining this project for about a year now, and my enthusiasm for it has dropped markedly in that time:

* the API absolutely cannot be salvaged. the primary APIs for loading image data require a buffer, with no length parameter for that buffer. without an externally-provided length, it is fundamentally unsafe to go parsing through said buffer. there is no way to recover from this save to change the API (ok, technically we could do something like capture SIGSEGV signals and handle page faults in userspace. i am quite unwilling to implement this personally, and would look askance at anyone who was).
  * even leaving this aside, the internals are rife with security problems -- see the continuous inflow of fuzzer-found bugs
* the internals use multiple distinct image decoding libraries, each with their own API and constrains, rather than a universal backend like OIIO or FFmpeg. this makes for tedious and unrewarding work.
* the library does not detect all properties of sixel support at runtime, and does not support terminal bitmap protocols beyond sixel, which is pretty much the worst of the lot (though most widely-implemented)

so no matter what, **we have to change the API** or this library is IMHO dead in the water. it's the image-processing equivalent of `gets()`. that's going to be a severe API+ABI break, affecting pretty much every user of the library. if we don't do it, though, i can no longer in good conscience recommend nor maintain libsixel.

so. the next question is the backend, and here i suggest another major change. in the past week, due to be released in 3.0.5, i've improved Notcurses' internal sixel encoder to the point where it's roughly equivalent or demonstrably superior to libsixel in all my test cases (prior to this work, Notcurses' sixel implementation was generally slower and inferior to that of libsixel). in particular, images from the AllRGB test set (4096x4096 using each RGB color once and only once) are handled by Notcurses in an order of magnitude less time (~1s vs ~10s+). in addition, of course, Notcurses probes support at runtime, drives all known terminal bitmap protocols, and uses FFmpeg/OIIO as a consolidated decoding backend.

I thus propose:

* We break the API as necessary (i'm pretty sure i have an issue and maybe even a PR on this) to de`gets()`-ize libsixel
* We reduce libsixel to a thin layer atop Notcurses (we'd need retain the dithering code, which NC doesn't support)

The two stand alone--it is not necessary to switch the backend in order to fix the frontend. With that said, I'm no longer very interested in maintaining the great spilled guts of libsixel, now that Notcurses IMHO eclipses it. I'm not, however, going to abuse my maintainer powers to unilaterally switch us, especially since there's a clear conflict of interest (I'm the author of Notcurses, for anyone unaware).

So what are your thoughts? Contributions from the public?

* is notcurses packaged everywhere that libsixel is?
* how much extra baggage does notcurses bring over libsixel?
* ought a disinterested party verify my claims of encoder superiority?

I do not consider the Notcurses switch necessary, but I do think it wise, and I would be disinclined to invest further work in libsixel otherwise.

--dank

**Comment by dankamongmen** [comment]
note that the APIs which accept a filename are ok; they need be updated to use the new internals, but other than that, they're good. again, i have an outstanding PR that does most of this (it was months ago; i'm not quite sure what its status was, but it's up).

**Comment by ghost** [comment]
Nice-to-haves:
* Arbitrary palette size with no limit, and no power-of-2 requirement.  Xterm supports 1024 colors.  Other terminals support many more (I've seen one report 65536 to the number of registers query).
* Creation of sixel images with transparency (omitted pixels), with an alpha cutoff parameter.
* No explicit dependencies on libpng, libjpeg, etc. in the encoder.  The command-line utilities of course will need those, but the encoder itself should not, it should be able to take a buffer of RGBA pixels and generate a string.
* An option to match the terminal background color (via OSC 4 query) for utilities like img2sixel.

**Reply by dankamongmen** [reply]
> * Arbitrary palette size with no limit, and no power-of-2 requirement.  Xterm supports 1024 colors.  Other terminals support many more (I've seen one report 65536 to the number of registers query).

yeah, contour answers with 65535 for sure. i have not tested above 1024, but i ought work fine. i do not have a power-of-2 requirement.

> * Creation of sixel images with transparency (omitted pixels), with an alpha cutoff parameter.

yep

> * No explicit dependencies on libpng, libjpeg, etc. in the encoder.  The command-line utilities of course will need those, but the encoder itself should not, it should be able to take a buffer of RGBA pixels and generate a string.

this requires a few notcurses calls; i would wrap it up behind a single libsixel call

> * An option to match the terminal background color (via OSC 4 query) for utilities like img2sixel.

oooh i like it. i've got all the pieces, but haven't put that together. hrmm, how best to express it...

**Comment by schrmh** [comment]
You might consider changing the name of the repo to "libsixel2" or — if you go the notcurses route — "libnotsixel" if it really breaks code of existing users.
I wonder if some of the affected code that saitoha wrote was actually written intentionally. Keep in mind, that he strived for backwards compatibility with real VTs and judging from his videos, he actually had access to a VT-330 and a VT-340.
Also when looking at some of his comments he made on issues, you can kind of see he was worried about that and given what projects he contributed too, it's kind of unlikely that he wasn't aware of some of the problems. E.g. in https://github.com/saitoha/libsixel/issues/44#issuecomment-120738824 he mentioned that he maintained a few terminal emulators and in the linked thread he mentioned ANSI X3.64/ECMA-48/ISO-6429 and ITU T.416.
Also, saitoha wrote most of the SIXEL patches for other terminals. By the way, I would say that this is actually the reason, why SIXEL is the de facto standard nowadays and not the barely supported IIP by iTerm or kitty.

**Reply by dankamongmen** [reply]
all valid points to be sure. i'd say the "libsixel2" move would be something based on the API changing, more than the backend.

we need that API change no matter what.

**Comment by dankamongmen** [comment]
so having looked at @hpjansson's libchafa, i think that's maybe a better backend? it is more closely aligned with libsixel's objectives, has a better sixel encoder than notcurses (for now), and has a smaller dep stack.

**Reply by j4james** [reply]
If libsixel is really unmaintainable, and you don't feel there's any value in contributing to this project anymore, why don't you just archive the repo and move over to libchafa instead? I don't understand why you'd want to rename another library as libsixel, regardless of whether it's chafa or notcurses. If I've understood you correctly, it's not even likely to be API compatible with libsixel, so it just seems like you're trying to take over the libsixel name.

**Reply by dankamongmen** [reply]
you raise a good point.

**Reply by ctrlcctrlv** [reply]
—yeah. I'd only feel comfortable doing that if we had some level of API—not even necessarily ABI, just C API—compatibility.

It will cause immense confusion downstream to rename something else libsixel.

No opposition to libsixel2, however.

**Reply by dankamongmen** [reply]
oh i never meant to suggest "rename something libsixel", only that libsixel be a thin compatibility wrapper around some other innards.

**Reply by ctrlcctrlv** [reply]
> If I've understood you correctly, it's not even likely to be API compatible with libsixel, so it just seems like you're trying to take over the libsixel name.

Ah, so @j4james did not indeed understand you correctly :-)

**Reply by j4james** [reply]
Quoting the original post that started this discussion: "the API absolutely cannot be salvaged". I don't know how else to interpret that other than to assume you'd be introducing a new API. So that means you'd be throwing out the original API, throwing out the internal implementation, and the only thing you seem interested in retaining is the name. Which part of that did I misunderstand?

I should add that I don't actually agree with the original premise stating that the API is fundamentally unsafe, but I don't care enough to argue that point. I just think if you don't like the way libsixel is implemented, and feel another library better meets your needs, then you should simply go ahead and use that other library.

If you aren't actually planning to use the libsixel name, and you're committing to leaving libsixel as it is, then I have no objections.

**Reply by ctrlcctrlv** [reply]
@j4james Obviously the two of us are different people, and so I can't answer for Nick. _It's my understanding_ that what he meant by this is that the API is too large, has some dangerous functions, has some badly implemented functions, but also, most importantly, tries to do too much.

I thought that API's will be dropped—perhaps even many at once, making two branches—but not _the entire API_.

**Reply by j4james** [reply]
@ctrlcctrlv Those are all great arguments for developing another competing library. But when you guys started this fork, I was under the impression it was about fixing bugs and merging new features that had been languishing in the PR backlog. What is being proposed now is an outright replacement of libsixel with another lib (maybe with a partially compatible API). That's something else entirely.

But you own the repo, so you can do what you want with it. It just feels very wrong to me. It's a bit like domain squatting using a well known company name to sell a competing product.

**Reply by ctrlcctrlv** [reply]
:shrug: Nothing is set in stone, thus the fact this discussion is happening at all. If we knew for sure what we wanted to do why discuss it? Yes, I wanted to merge a lot of PR's—which I did. Pushing things forward is probably going to involve some level of deprecations. I don't yet know what level.

**Reply by dankamongmen** [reply]
i think i have been unclear, which is of course my failing and fault.

allow me to summarize:

1) The current *CORE API* of this project cannot be salvaged. For purposes of discussion, let us reduce the "core API" to a single function, one which does not exist (but is representative):

 `libsixel_display(libsixel* context, const void* data)`

This API is broken in the same way that `gets()` is broken: there is no way to know how long `data` is. It is unsafe to use. If this is not obvious, I'm not sure what to say.

2) Making this usable would require an API change. Either a length parameter would need be added to the function above (ala `fgets()`), or a call would need be added which provided the length, and would need be invoked prior to use of this function. Again, I don't really see how anyone can argue with this.

3) *ORTHOGONAL TO THIS*, i proposed changing the *backend*. It was my mistake to introduce these two concepts in the same discussion, and I apologize for any confusion. This is *not necessary* for resolving the problem above. It *is necessary* for *my continued involvement in the project*, but I am of no importance.

At no time did I conceive of (or advocate for, so far as I can tell) any largescale change to the libsixel API.

In any case, I no longer wish to spend time on this project. Any disruption to the current API will require changing dozens of calling users, and quotes like @j4james's "It just feels very wrong to me. It's a bit like domain squatting using a well known company name to sell a competing product" aren't something I care to deal with. In my opinion, users of this library fucked up when they used and tolerated what was obviously a fundamentally broken API. If you're writing C, and passing buffers to users of those buffers without passing a buffer length, you probably shouldn't be writing C.

I have my own projects, and this is a distraction. Furthermore, at least one of those projects (Notcurses) can be seen as a competitor to this project, and thus my continued involvement can be seen as a conflict of interest. I thus recuse myself from this fork, and abandon my proposed changes to the API. I release any relevant comments and code of mine to the public domain, to be used if they are deemed useful. I wish the project the best going forward.

**Reply by dankamongmen** [reply]
https://github.com/libsixel/libsixel/commit/2855eea71effe4d09890f4633f7dd568dbdfae3f good luck!

**Reply by j4james** [reply]
First off I must apologize for suggesting you're using the libsixel name to promote your own library. I know that wasn't your intention, but you have to understand that that is exactly what it looks like from the outside. It gives a terrible impression and it's so unnecessary.

If you want to make it easy for libsixel users to migrate to another library like notcurses or chafa, there's no reason why you can't produce a wrapper lib that emulates the libsixel API, and include that as part of the notcurses project (and the same thing for chafa). And if you really think this codebase should be abandoned, you could always put something in the README saying you've decided not to maintain it, and provide links to alternatives like notcurses and chafa.

None of that requires erasing all of saitoha's work, though, and if someone else wanted to continue maintaining this lib in its current form (possibly even saitoha returning one day), the option is at least still there for them.

> `libsixel_display(libsixel* context, const void* data)`
> 
> This API is broken in the same way that `gets()` is broken

I wish you'd chosen a real example from the lib for this discussion, because this just seems like a strawman argument. If there really is a function like `gets` in the lib then I would agree that it's unsafe. But the bug report that I think triggered this whole crisis (issue #46), was not actually a valid complaint in my opinion, and should have been closed as "wontfix".

If I write code that produces a buffer overflow like this:

    unsigned char data[4096];
    fread(data, 400, 300, fp);

I don't then get to complain that there's a critical vulnerability in the `fread` function. That's my damn fault for passing it invalid parameters. And that's almost exactly what was reported in issue #46, which looked something like this:

    unsigned char data[4096];
    sixel_encoder_encode_bytes(encoder,(unsigned char*)data,400,300,SIXEL_PIXELFORMAT_RGBA8888,NULL,(-1));

If you expect a library to be immune to callers passing in bogus parameters like that, then C is definitely not the language for you. And adding an additional length parameter doesn't really help anything, because you can just as easily produce an overflow by setting the length incorrectly.

I didn't look at your proposed PR in much detail, so maybe I've missed something, but the impression I got was that all of the APIs you were patching were of a similar form. They already had an implicit length parameter (in the sense of width * height * size_of_pixelformat), so there was no compelling need for an explicit length as far as I could see.

Something like `gets` is completely different, since the caller has no way of knowing how big to make the buffer. But that's not the case for these libsixel functions - you know exactly how big the buffer should be. If you don't make it big enough (whether intentionally or otherwise), that's a bug in *your* code - not the library.

> I have my own projects, and this is a distraction. [...] I thus recuse myself from this fork, and abandon my proposed changes to the API. 

I think you're making the right choice here. You obviously weren't happy working on this code base, and there's no need to make yourself miserable maintaining someone else's code when there are better things you could be working on.

**Reply by dankamongmen** [reply]
no apologies necessary, @j4james -- i agree completely with your comment on how it looked from the outside. to the degree my reply was strained, it was due to fluster and embarrassment, not offense. as always, your insightful comments are appreciated.

i disagree with some of your comments, though not fundamentally. let me back off to this: without the proposed API change, there will be an unceasing inflow of fuzzing-based "security" reports, whether they're platonic security vulnerabilities or not. i think we can agree on that? not that it particularly matters =].

no matter what, shifting all callers was unlikely to happen.

i wish all involved the very best! <3 --dank

**Reply by hpjansson** [reply]
I dunno about total length, but a `rowstride` parameter can be useful, both for more explicit bounding (no need to look up size_of_pixelformat) and for cases where you have a buffer with row padding. You can also efficiently pass in a sub-rectangle of a bigger image that way.

Thanks for the confidence re. libchafa. I think it'd be possible to scrape the sixel encoder out of that and turn it into a public domain .c/.h blob that could be copied into other projects. Might be more palatable to tmux, screen, etc. But Copious Spare Time and all that :)

And a big thanks to @saitoha for kicking off the ongoing sixel revival.
